tools/wake-unit/doc.cpp

```diff
@@ -0,0 +1,378 @@
+  }
+
+  wcl::doc d = std::move(builder).build();
+  ASSERT_EQUAL(3000000u, d.character_count());
+}
+
+TEST(doc_undo) {
+  wcl::doc_builder builder;
+  builder.append("Hello");
+  builder.append(" ");
+  builder.append("World");
+  builder.append("!");
+
+  builder.undo();
+  builder.undo();
+
+  wcl::doc d = std::move(builder).build();
+  std::string expected = "Hello ";
+
+  EXPECT_EQUAL(expected.size(), d.character_count());
+  EXPECT_EQUAL(expected, d.as_string());
+}
+
+TEST(doc_geometry) {
+  {
+    wcl::doc_builder builder;
+    builder.append("Hello");
+
+    EXPECT_EQUAL(5u, builder.first_width());
+    EXPECT_EQUAL(5u, builder.last_width());
+    EXPECT_EQUAL(5u, builder.max_width());
+    EXPECT_EQUAL(0u, builder.newline_count());
+    EXPECT_EQUAL(1u, builder.height());
+
+    wcl::doc d = std::move(builder).build();
+
+    EXPECT_EQUAL(5u, d.first_width());
+    EXPECT_EQUAL(5u, d.last_width());
+    EXPECT_EQUAL(5u, d.max_width());
+    EXPECT_EQUAL(0u, d.newline_count());
+    EXPECT_EQUAL(1u, d.height());
+  }
+
+  {
+    wcl::doc_builder builder;
+    builder.append("Hello\n");
+
+    EXPECT_EQUAL(5u, builder.first_width());
+    EXPECT_EQUAL(0u, builder.last_width());
+    EXPECT_EQUAL(5u, builder.max_width());
+    EXPECT_EQUAL(1u, builder.newline_count());
+    EXPECT_EQUAL(2u, builder.height());
+
+    wcl::doc d = std::move(builder).build();
+
+    EXPECT_EQUAL(5u, d.first_width());
+    EXPECT_EQUAL(0u, d.last_width());
+    EXPECT_EQUAL(5u, d.max_width());
+    EXPECT_EQUAL(1u, d.newline_count());
+    EXPECT_EQUAL(2u, d.height());
+  }
+
+  {
+    wcl::doc_builder builder;
+    builder.append("Hello\n");
+    builder.append("World!");
+
+    EXPECT_EQUAL(5u, builder.first_width());
+    EXPECT_EQUAL(6u, builder.last_width());
+    EXPECT_EQUAL(6u, builder.max_width());
+    EXPECT_EQUAL(1u, builder.newline_count());
+    EXPECT_EQUAL(2u, builder.height());
+
+    wcl::doc d = std::move(builder).build();
+
+    EXPECT_EQUAL(5u, d.first_width());
+    EXPECT_EQUAL(6u, d.last_width());
+    EXPECT_EQUAL(6u, d.max_width());
+    EXPECT_EQUAL(1u, d.newline_count());
+    EXPECT_EQUAL(2u, d.height());
+  }
+
+  {
+    wcl::doc_builder builder;
+    builder.append("Hello");
+    builder.append("\nWorld!");
+
+    EXPECT_EQUAL(5u, builder.first_width());
+    EXPECT_EQUAL(6u, builder.last_width());
+    EXPECT_EQUAL(6u, builder.max_width());
+    EXPECT_EQUAL(1u, builder.newline_count());
+    EXPECT_EQUAL(2u, builder.height());
+
+    wcl::doc d = std::move(builder).build();
+
+    EXPECT_EQUAL(5u, d.first_width());
+    EXPECT_EQUAL(6u, d.last_width());
+    EXPECT_EQUAL(6u, d.max_width());
+    EXPECT_EQUAL(1u, d.newline_count());
+    EXPECT_EQUAL(2u, d.height());
+  }
+
+  {
+    wcl::doc_builder builder;
+    builder.append("He\nllo");
+    builder.append("Worl\nd!");
+
+    EXPECT_EQUAL(2u, builder.first_width());
+    EXPECT_EQUAL(2u, builder.last_width());
+    EXPECT_EQUAL(7u, builder.max_width());
```

Comment:
yep here's the test. looks like your merge without my approval gets my seal of approval ;)